prov/verbs/src/verbs_mr.c

```diff
@@ -130,37 +154,292 @@ fi_ibv_mr_reg(struct fid *fid, const void *buf, size_t len,
+		.iov_base	= buf,
+		.iov_len	= len,
+	};
+	struct fi_mr_attr attr = {
+		.mr_iov		= &iov,
+		.iov_count	= 1,
+		.access		= access,
+	};
+	struct fi_ibv_mem_desc *mdesc =
+		fi_ibv_mr_common_cache_reg(domain, &attr);
+	if (OFI_UNLIKELY(!mdesc))
+		return -FI_EAVAIL;
+	*md = *mdesc;
+	md->len = len;
+	return FI_SUCCESS;
+}
+
+static inline
+int fi_ibv_mr_internal_cache_dereg(struct fi_ibv_mem_desc *md)
+{
+	fi_ibv_common_cache_dereg(md);
+	md->mr = NULL;
+	return FI_SUCCESS;
+}
+
+static inline
+int fi_ibv_mr_internal_ex_reg(struct fi_ibv_domain *domain, void *buf,
+			      size_t len, uint64_t access,
+			      struct fi_ibv_mem_desc *md)
+{
+	return ((len >= fi_ibv_gl_data.mr_cache_lazy_size) ?
+		fi_ibv_mr_internal_cache_reg(domain, buf, len, access, md) :
+		fi_ibv_mr_internal_reg(domain, buf, len, access, md));
+}
+
+static inline
+int fi_ibv_mr_internal_ex_dereg(struct fi_ibv_mem_desc *md)
+{
+	return ((md->len >= fi_ibv_gl_data.mr_cache_lazy_size) ?
+		fi_ibv_mr_internal_cache_dereg(md) :
+		fi_ibv_mr_internal_dereg(md));
 }
 
 static int fi_ibv_mr_regattr(struct fid *fid, const struct fi_mr_attr *attr,
-		uint64_t flags, struct fid_mr **mr)
+			     uint64_t flags, struct fid_mr **mr)
+{
+	struct fi_ibv_mem_desc *md;
+	int ret;
+
+	ret = fi_ibv_mr_regattr_check_args(fid, attr, flags);
+	if (OFI_UNLIKELY(ret))
+		return ret;
+
+	md = calloc(1, sizeof(*md));
+	if (OFI_UNLIKELY(!md))
+		return -FI_ENOMEM;
+
+	md->domain = container_of(fid, struct fi_ibv_domain,
+				  util_domain.domain_fid.fid);
+	md->mr_fid.fid.ops = &fi_ibv_mr_ops;
+
+	ret = fi_ibv_mr_reg_common(md, fi_ibv_mr_ofi2ibv_access(attr->access,
+								md->domain),
+				   attr->mr_iov[0].iov_base,
+				   attr->mr_iov[0].iov_len, attr->context);
+	if (OFI_UNLIKELY(ret))
+		goto err;
+
+	*mr = &md->mr_fid;
+	return FI_SUCCESS;
+err:
+	free(md);
+	return ret;
+}
+
+FI_IBV_DEFINE_MR_REG_OPS(_)
+
+static int fi_ibv_mr_cache_close(fid_t fid)
 {
-	return fi_ibv_mr_regv(fid, attr->mr_iov, attr->iov_count, attr->access,
-			0, attr->requested_key, flags, mr, attr->context);
+	struct fi_ibv_mem_desc *md =
+		container_of(fid, struct fi_ibv_mem_desc, mr_fid.fid);
+	
+	fi_ibv_common_cache_dereg(md);
+
+	return FI_SUCCESS;
 }
 
-struct fi_ops_mr fi_ibv_domain_mr_ops = {
-	.size = sizeof(struct fi_ops_mr),
-	.reg = fi_ibv_mr_reg,
-	.regv = fi_ibv_mr_regv,
-	.regattr = fi_ibv_mr_regattr,
+static struct fi_ops fi_ibv_mr_cache_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = fi_ibv_mr_cache_close,
+	.bind = fi_no_bind,
+	.control = fi_no_control,
+	.ops_open = fi_no_ops_open,
 };
+
+int fi_ibv_monitor_subscribe(struct ofi_mem_monitor *notifier, void *addr,
+			     size_t len, struct ofi_subscription *subscription)
+{
+	return FI_SUCCESS;
+}
+
+void fi_ibv_monitor_unsubscribe(struct ofi_mem_monitor *notifier, void *addr,
+				size_t len, struct ofi_subscription *subscription)
+{
+}
+
+struct ofi_subscription *fi_ibv_monitor_get_event(struct ofi_mem_monitor *notifier)
+{
+	return NULL;
+}
```

Comment:
These are fine for now as placeholders, but I think the final solution (guessing it will try to hook alloc/free) should put the monitor implementation into util code.

Reply:
yes, sure. the final solution will do that